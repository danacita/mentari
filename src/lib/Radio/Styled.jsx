import styled, { css } from 'styled-components';

import Color from '../Styles/bases/Color';

const RadioGroup = styled.div`
  ${props =>
    props.slideOption
      ? css`
          display: inline-block;
          border: 1px solid ${Color.grey191};
          border-radius: 10px;
          > div {
            margin-right: 0;
            > label {
              padding: 4px 8px;
            }
            > label::before {
              display: none;
            }
            > label::after {
              width: 100% !important;
              height: 100% !important;
              border-radius: 10px !important;
              left: 0 !important;
              background-color: transparent !important;
              border: 2px solid ${Color.darkBlue};
            }
          }
        `
      : css`
          display: block;
        `};
`;

const RadioWrapper = styled.div`
  font-size: 14px;
  display: inline-block;
  margin-right: 16px;
  &:last-child {
    margin-right: 0;
  }
`;

const RadioInput = styled.input`
  position: absolute;
  opacity: 0;
  &:focus {
    & + label {
      &:before {
        border: 1px solid;
        border-color: ${Color.blue};
      }
    }
  }
  & + label {
    cursor: pointer;
    position: relative;
    display: flex;
    align-items: center;
    font-weight: 400;
    margin-bottom: 0;
    &:before {
      content: '';
      margin-right: 8px;
      display: inline-block;
      vertical-align: middle;
      width: 18px;
      height: 18px;
      background-color: ${Color.white};
      border: 1px solid ${Color.grey191};
      border-radius: 100%;
    }
  }
  &:checked + label {
    &:before {
      background-color: ${Color.white};
      border-color: ${Color.grey191};
    }
    &:after {
      content: '';
      position: absolute;
      width: 12px;
      height: 12px;
      background-color: ${Color.orange};
      border-radius: 100%;
      left: 3px;
      ${props =>
        props.slideOption
          ? css`
              top: 0;
            `
          : css`
              top: 3px;
            `}
    }
  }
`;

export { RadioWrapper, RadioInput, RadioGroup };
