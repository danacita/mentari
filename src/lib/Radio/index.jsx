/* eslint-disable jsx-a11y/label-has-for */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { RadioWrapper, RadioInput, RadioGroup } from './Styled';

const propTypes = {
  fieldData: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      name: PropTypes.string
    })
  ).isRequired,
  name: PropTypes.string.isRequired,
  defaultValue: PropTypes.string,
  withRedux: PropTypes.bool,
  block: PropTypes.bool,
  customOnChange: PropTypes.func,
  onChange: PropTypes.func,
  slideOption: PropTypes.bool,
  // input prop is passed by redux-form to keep field in sync with state
  input: PropTypes.shape({
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.bool]),
    onChange: PropTypes.func
  })
};

const defaultProps = {
  customOnChange: () => {},
  onChange: () => {},
  defaultValue: '',
  slideOption: false,
  withRedux: true,
  block: false,
  input: {}
};

const Radio = ({
  fieldData,
  input,
  customOnChange,
  slideOption,
  withRedux,
  onChange,
  name,
  block,
  defaultValue
}) => (
  <Fragment>
    {withRedux ? (
      <RadioGroup slideOption={slideOption}>
        {fieldData.map(option => (
          <RadioWrapper key={option.id} block={block}>
            <RadioInput
              id={option.id}
              data-testid={option.id}
              name={input.name}
              type="radio"
              block={block}
              value={option.id}
              checked={input && input.value === option.id}
              slideOption={slideOption}
              onChange={value => {
                customOnChange(value);
                input.onChange(value);
              }}
            />
            <label htmlFor={option.id}>{option.name}</label>
          </RadioWrapper>
        ))}
      </RadioGroup>
    ) : (
      <RadioGroup slideOption={slideOption}>
        {fieldData.map(option => (
          <RadioWrapper key={option.id} block={block}>
            <RadioInput
              id={option.id}
              data-testid={option.id}
              name={name}
              type="radio"
              value={option.id}
              slideOption={slideOption}
              onChange={onChange}
              checked={defaultValue === option.id}
            />
            <label htmlFor={option.id}>{option.name}</label>
          </RadioWrapper>
        ))}
      </RadioGroup>
    )}
  </Fragment>
);

Radio.propTypes = propTypes;
Radio.defaultProps = defaultProps;

export default Radio;
