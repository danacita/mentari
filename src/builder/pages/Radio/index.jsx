import React, { Component, Fragment } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { atomOneDarkReasonable } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import { ContentWithRightBar } from '../Styled';
import { Margin, Radio, Checkbox } from '../../../lib';
import PropsShared from '../../shared/Props';
import DemoShared from '../../shared/Demo';
import CodeShared from '../../shared/Code';
import RightbarShared from '../../shared/Rightbar';

class RadioPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      radioDefault: 'one',
      radioBlock: false,
      radioSlide: false
    };
  }

  render() {
    const { radioDefault, radioBlock, radioSlide } = this.state;
    const codeString = `import { Radio } from '@danacita/mentari';\n<Radio fieldData={[{ id: 'one', name: 'Option one' }, { id: 'two', name: 'Option two' }, { id: 'three', name: 'Option three' }, { id: 'four', name: 'Option four' }]} ${
      radioBlock ? 'block ' : ''
    }${radioSlide ? 'slideOption' : ''}/>`;
    return (
      <Fragment>
        <ContentWithRightBar>
          <h1>Radio</h1>
          <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
          <Margin top={24}>
            <PropsShared>
              <li>
                <span>id</span>
                <span>Apply a id to the control element</span>
              </li>
              <li>
                <span>name</span>
                <span>Set the name of Radio</span>
              </li>
              <li>
                <span>name</span>
                <span>Set the name of Radio</span>
              </li>
              <li>
                <span>slideOption</span>
                <span>Apply the options to slide</span>
              </li>
              <li>
                <span>block</span>
                <span>Apply the full width of a parent</span>
              </li>
            </PropsShared>
          </Margin>
          <DemoShared>
            <h4>Demo</h4>
            <p>You can change the props via the component builder on your right</p>
            <Margin top={16}>
              <Radio
                id="radioDemo"
                name="radioDemo"
                fieldData={[
                  { id: 'one', name: 'Option one' },
                  { id: 'two', name: 'Option two' },
                  { id: 'three', name: 'Option three' },
                  { id: 'four', name: 'Option four' }
                ]}
                withRedux={false}
                defaultValue={radioDefault}
                onChange={e => this.setState({ radioDefault: e.target.id })}
                block={radioBlock}
                slideOption={radioSlide}
              />
            </Margin>
          </DemoShared>
          <h4>Code</h4>
          <Margin top={16}>
            <CodeShared copyCode={codeString}>
              <SyntaxHighlighter language="jsx" style={atomOneDarkReasonable} showLineNumbers>
                {codeString}
              </SyntaxHighlighter>
            </CodeShared>
          </Margin>
          <RightbarShared title="Radio">
            <Margin bottom={16}>
              <h4>State of Radio</h4>
            </Margin>
            <Margin bottom={8}>
              <Checkbox
                id="radioBlock"
                name="radioBlock"
                label="Block"
                checked={radioBlock}
                withRedux={false}
                onChange={e => this.setState({ radioBlock: e.target.checked })}
              />
            </Margin>
            <Margin bottom={24}>
              <Checkbox
                id="radioSlide"
                name="radioSlide"
                label="Slide"
                checked={radioSlide}
                withRedux={false}
                onChange={e => this.setState({ radioSlide: e.target.checked })}
              />
            </Margin>
          </RightbarShared>
        </ContentWithRightBar>
      </Fragment>
    );
  }
}

export default RadioPage;
