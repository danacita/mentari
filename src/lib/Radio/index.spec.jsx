import React from 'react';
import renderer from 'react-test-renderer';

import Radio from '.';

describe('Radio', () => {
  it('renders correctly', () => {
    const fieldData = [
      {
        id: 1,
        name: 'Test Name'
      }
    ];

    const input = {
      value: 'Value',
      onChange: () => {}
    };

    const tree = renderer.create(<Radio fieldData={fieldData} input={input} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
