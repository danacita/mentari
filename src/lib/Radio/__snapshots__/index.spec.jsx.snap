// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<div
  className="sc-bdVaJa jCNFIm"
>
  <div
    className="sc-bwzfXH hCBhkM"
  >
    <input
      checked={false}
      className="sc-htpNat gAWSad"
      data-testid={1}
      id={1}
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      htmlFor={1}
    >
      Test Name
    </label>
  </div>
</div>
`;
